### PR TITLE
Apply clang-tidy autofixes

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This changeset is large, but it's not very substantial. It's all the automated fixes produced by clang-tidy using our script. The bulk of the changes are either adding `[[nodiscard]]` to many functions or changing const ref args to pass by value and then move in cases where the parameter is only used to set a value. There are also some places where clang-tidy preferred either more or less namespacing of objects depending on the current namespace. The goal is to enable clang-tidy in CI, which we made progress towards in #9860 but stalled in #10064. This PR contains the first set of changes that will required for such a check to pass.

I've marked this PR as breaking because some of the functions now marked as `[[nodiscard]]` are public APIs, so if consumers were ignoring the return values they will now see warnings, and if they are compiling with warnings as errors then the builds will break.

Contributes to #584 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/detail/aggregation/aggregation.hpp
Comment (bdice) [inline cpp/include/cudf/detail/aggregation/aggregation.hpp:876]: Should this take a `string_view` instead of a `string` that we move? This happens quite a few times in this PR. I understand the potential (small?) benefit of move-semantics here, but it feels more idiomatic to accept a `string_view`.

File: cpp/include/cudf/detail/structs/utilities.hpp
Comment (bdice) [inline cpp/include/cudf/detail/structs/utilities.hpp:175]: Removing explicit namespaces feels like a loss of information that is helpful to the reader. Can these be reverted?

File: cpp/include/cudf/interop.hpp
Comment (bdice) [inline cpp/include/cudf/interop.hpp:126]: Another case where maybe we want to accept a `string_view`.

File: cpp/include/cudf/io/types.hpp
Comment (bdice) [inline cpp/include/cudf/io/types.hpp:251]: `string_view`?

File: cpp/include/cudf/join.hpp
Comment (bdice) [inline cpp/include/cudf/join.hpp:386]: Seems like a typo.
```suggestion
   * the result of performing a full join between two tables with `build` and `probe`
```

File: cpp/include/cudf/utilities/span.hpp
Comment (bdice) [inline cpp/include/cudf/utilities/span.hpp:372]: We don't need to use move semantics on a primitive (int) type.

File: cpp/src/io/parquet/predicate_pushdown.cpp
Comment (bdice) [inline cpp/src/io/parquet/predicate_pushdown.cpp:477]:
```suggestion
  for (auto const& input_row_group_index : input_row_group_indices) {
```

File: cpp/src/io/parquet/predicate_pushdown.cpp
Comment (bdice) [inline cpp/src/io/parquet/predicate_pushdown.cpp:479]:
```suggestion
    for (auto const rg_idx : input_row_group_index) {
```

File: cpp/src/io/parquet/reader_impl_helpers.cpp
Comment (bdice) [inline cpp/src/io/parquet/reader_impl_helpers.cpp:1094]: Lots of west-const in this PR... can we autoformat to east const to match the rest of the library? (We don't have this option enabled in clang-format because it's supposedly bug-prone, but I don't think we observed any actual bugs when we last ran a manual formatting pass.)

**Review by bdice — COMMENTED** [review]
A few initial comments -- I got through 55 / 235 files and decided to share my comments before going further, since some apply throughout the PR.

File: cpp/include/cudf/detail/aggregation/aggregation.hpp
Comment (vyasr) [reply cpp/include/cudf/detail/aggregation/aggregation.hpp:876]: I agree, but I'm not sure I want to attempt that in this PR. Would you be OK with opening an issue to replace `std::string` parameters with `std::string_view` and doing it in a follow-up?

File: cpp/include/cudf/detail/structs/utilities.hpp
Comment (vyasr) [reply cpp/include/cudf/detail/structs/utilities.hpp:175]: To be honest I didn't take the time to understand the rules clang-tidy followed for when it liked namespaces and when it didn't. Let me see what I can find about that.

File: cpp/include/cudf/join.hpp
Comment (vyasr) [reply cpp/include/cudf/join.hpp:386]: Ugh yeah this happened a few times. I thought I had caught all of these, but I guess not. I'll do another pass, thanks.

File: cpp/src/io/parquet/reader_impl_helpers.cpp
Comment (vyasr) [reply cpp/src/io/parquet/reader_impl_helpers.cpp:1094]: Hmm yeah at minimum I can do a manual pass on this PR and fix anything that goes wrong.

Comment (vyasr) [reply cpp/src/io/parquet/reader_impl_helpers.cpp:1094]: Manual pass worked fine. [The warning is still present in the clang-format docs](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#qualifieralignment), though, so I've left the option out of our configuration.

File: cpp/src/io/parquet/predicate_pushdown.cpp
Comment (vyasr) [reply cpp/src/io/parquet/predicate_pushdown.cpp:477]: Good catch!

File: cpp/include/cudf/utilities/span.hpp
Comment (vyasr) [reply cpp/include/cudf/utilities/span.hpp:372]: Checked the rest with
```
(rapids) coder _ ~/cudf $ git diff upstream/branch-24.08 | grep -E 'std::move' | grep -E '^[+-]'
+      _source{std::move(user_defined_aggregator)},
+    : _flattened_columns{std::move(flattened_columns_)},
+  column_metadata(std::string _name) : name(std::move(_name)) {}
+    : arrow_file(std::move(file))
+    : _sink(std::move(sink)), _table(table), _rows_per_chunk(table.num_rows())
+    : _sink(std::move(sink)), _table(std::move(table)), _rows_per_chunk(table.num_rows())
+    : _sink(std::move(sink)), _table(std::move(table))
+  chunked_orc_writer_options(sink_info sink) : _sink(std::move(sink)) {}
+    : _sink(std::move(sink)), _table(std::move(table))
+  explicit chunked_parquet_writer_options(sink_info sink) : _sink(std::move(sink)) {}
+    : name(std::move(_name)), is_nullable(_is_nullable), is_binary(_is_binary)
+    : schema_idx(_schema_idx), name(std::move(_name)), has_repetition(_has_repetition)
+    : arrow_buffer(std::move(arrow_buffer))
+  : ptx(std::move(ptx_)),
+    function_name(std::move(function_name_)),
+    output_arg_type(std::move(output_arg_type_)),
+    : col(std::move(col)), output_dtype(output_dtype), init(init), mr(mr), stream(stream)
-    src.push_back(cudf::test::structs_column_wrapper(std::move(inputs)));
+    src.emplace_back(std::move(inputs));
-      inner_structs.push_back(cudf::test::structs_column_wrapper(std::move(children)));
+      inner_structs.emplace_back(std::move(children));
-    src.push_back(cudf::test::structs_column_wrapper(std::move(inputs)));
+    src.emplace_back(std::move(inputs));
-      inner_structs.push_back(cudf::test::structs_column_wrapper(std::move(children)));
+      inner_structs.emplace_back(std::move(children));
-    return structs{std::move(expected_vector_of_columns), {0, 1, 1, 1, 1}};
+    return structs{std::move(expected_vector_of_columns), {false, true, true, true, true}};
-    cudf::test::structs_column_wrapper{{std::move(struct_children)}, {0, 0, 1, 1, 1, 1}};
+  auto outer_struct = cudf::test::structs_column_wrapper{{std::move(struct_children)},
+    : gby_column(std::move(gby)),
+      oby_column(std::move(oby)),
+      agg_column(std::move(agg)),
+      preceding(std::move(preceding_scalar)),
+      following(std::move(following_scalar)),
-  cudf::test::structs_column_wrapper valid_struct_col(std::move(src_columns), {1});
+  cudf::test::structs_column_wrapper valid_struct_col(std::move(src_columns), {true});
-  cudf::test::structs_column_wrapper invalid_struct_col(std::move(src_columns), {0});
+  cudf::test::structs_column_wrapper invalid_struct_col(std::move(src_columns), {false});
```
I don't see any other moves that are similarly superfluous.

File: cpp/include/cudf/detail/structs/utilities.hpp
Comment (vyasr) [reply cpp/include/cudf/detail/structs/utilities.hpp:175]: Wait hold on, this is adding explicit namespaces, not removing them. Did you like it better with fewer namespaces?

File: cpp/include/cudf/utilities/span.hpp
Comment (vyasr) [reply cpp/include/cudf/utilities/span.hpp:372]: I don't know how to tell clang-tidy that these moves aren't necessary. I can add a NOLINT comment to this line, but [the rule doesn't appear to be customizable in any way](https://clang.llvm.org/extra/clang-tidy/checks/modernize/pass-by-value.html). Let me know what you'd like to do here.

**Comment by vyasr** [comment]
I'm not sure what the right thing to do is for generated and vendored files. There are two in particular that are a bit problematic, [cxxopts.hpp](https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/include/cudf_test/cxxopts.hpp) and [Schema_generated.h](https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/src/io/parquet/ipc/Schema_generated.h). The problem is that the way clang-tidy works is recursive, i.e. you point it to a source file and it checks all headers included by that file. As a result, it's not sufficient to use something like pre-commit filtering of files beforehand to tell it what files to ignore. The way to specify what files to include is using the `-header-filter` argument to clang-tidy itself. However, [this argument does not support lookahead/lookbehind in regexes](https://stackoverflow.com/questions/71797349/is-it-possible-to-ignore-a-header-with-clang-tidy), which is the only way I can imagine filtering out a specific set of files right now. The as-yet-unreleased version 19 includes a new feature to exclude headers, though. Our options are to either go ahead and apply changes to these files, or to wait until clang-tidy 19 is released and update to using that and assume that the new option will allow us to do the necessary filtering.

I should note that while we do currently have pre-commit filtering out the cxxopts.hpp file, the Schema_generated.h file has not been filtered, so we are already applying clang-format to it. That being the case, I would be fine with just removing the filters and applying clang-tidy and clang-format uniformly to both files, but I wanted to see what others thought first.

**Comment by davidwendt** [comment]
> I should note that while we do currently have pre-commit filtering out the cxxopts.hpp file, the Schema_generated.h file has not been filtered, so we are already applying clang-format to it. That being the case, I would be fine with just removing the filters and applying clang-tidy and clang-format uniformly to both files, but I wanted to see what others thought first.

I'd rather add the `Schema_generated.h` to the pre-commit filter if that is possible. 
I'm ok with updating the files manually using this PR and we can revisit again once version 19 is out.

I'm more concerned that we have some cpp-tests failing with these changes.

**Comment by vyasr** [comment]
@davidwendt apologies, I had fixed the failing test but just realized I was pushing to the wrong branch. It was just an issue of a missed NOLINT; clang-tidy doesn't understand all of the extra braces that we sometimes need in tests in order to make things like nested list columns work.

As I mentioned, we can't use a pre-commit filter:
> The problem is that the way clang-tidy works is recursive, i.e. you point it to a source file and it checks all headers included by that file. As a result, it's not sufficient to use something like pre-commit filtering of files beforehand to tell it what files to ignore.

It's not really about pre-commit, it's about clang-tidy itself. If you run clang-tidy on any source file, it will also detect errors in any of the headers that file includes.

**Comment by davidwendt** [comment]
Ok, this part sounded like we have a mistake. I'm proposing we fix this:
> I should note that while we do currently have pre-commit filtering out the cxxopts.hpp file, the Schema_generated.h file has not been filtered, so we are already applying clang-format to it.

And this part sounded like we can solve the clang-tidy issue with version 19:
> The as-yet-unreleased version 19 includes a new feature to exclude headers, though.

I'm proposing we merge this PR (once it passes) and wait for 19.

**Comment by vyasr** [comment]
Got it. Yes, I can fix the exclusion for clang-format for Schema_generated.h

As far as waiting, [I'm observing a biannual release cadence](https://releases.llvm.org/) in March and September. I would (admittedly selfishly) like to get this work done before then personally, but I understand the desire to not have that behavior, so I'll defer to you on that if you think it's worth avoiding the modifications to the generated file.

File: cpp/src/io/parquet/reader_impl_helpers.cpp
Comment (bdice) [reply cpp/src/io/parquet/reader_impl_helpers.cpp:1094]: I dug up the original reviews for the PR where this feature and that warning were added. https://reviews.llvm.org/D69764#inline-1055542

Basically the breaking cases are things like:

```cpp
#define INTPTR int *

const INTPTR a;
```

where a macro would mess up the semantics of `const` and `*` by swapping the order.

I'd be okay with accepting that edge case and enabling this setting across the code base.

File: cpp/include/cudf/utilities/span.hpp
Comment (bdice) [reply cpp/include/cudf/utilities/span.hpp]: I am happy with the analysis you did. I'd accept this as-is and move on. You could add NOLINT if you want, either way is fine.

File: cpp/src/io/parquet/predicate_pushdown.cpp
Comment (bdice) [inline cpp/src/io/parquet/predicate_pushdown.cpp:477]:
```suggestion
  for (auto const& input_row_group_index : input_row_group_indices) {
```

File: cpp/src/io/parquet/predicate_pushdown.cpp
Comment (bdice) [inline cpp/src/io/parquet/predicate_pushdown.cpp:479]:
```suggestion
    for (auto const rg_idx : input_row_group_index) {
```

File: cpp/include/cudf/detail/structs/utilities.hpp
Comment (bdice) [reply cpp/include/cudf/detail/structs/utilities.hpp:175]: 😅 I might have read this wrong. I like using explicit namespaces!

File: cpp/include/cudf/detail/aggregation/aggregation.hpp
Comment (bdice) [reply cpp/include/cudf/detail/aggregation/aggregation.hpp:876]: That sounds fine to me. Totally happy to defer on this.

**Review by bdice — APPROVED** [review]
I have read enough of this that I think it can be merged. I didn't read the changes to `cpp/tests` but I read everything else.

Aside from the other comments I just left, I would like one more spot-check for things like `indices` -> `indice` in the for-loop simplifications. I'm not sure how to check for those. Maybe grep the diff for `for (.*:.*` or something like that.

File: cpp/include/cudf/detail/aggregation/aggregation.hpp
Comment (vyasr) [reply cpp/include/cudf/detail/aggregation/aggregation.hpp:876]: https://github.com/rapidsai/cudf/issues/15907

File: cpp/src/io/parquet/reader_impl_helpers.cpp
Comment (vyasr) [reply cpp/src/io/parquet/reader_impl_helpers.cpp:1094]: OK, I'll add it for now. We can always disable it in the future if developers discover that it breaks things for them.

**Comment by vyasr** [comment]
I did the spot-check and didn't find anything. Thanks for being so diligent!

I'll do a follow-up PR where I fix the clang-tidy errors that don't get autofixed. I figured this changeset was large enough as is.

File: cpp/include/cudf/scalar/scalar.hpp
Comment (karthikeyann) [inline cpp/include/cudf/scalar/scalar.hpp:570]: Question: This is interesting. Why is override required here?

File: cpp/include/cudf/scalar/scalar_factories.hpp
Comment (karthikeyann) [inline cpp/include/cudf/scalar/scalar_factories.hpp:81]: There are few places where `cudf::column` is used instead of `column`. At someplaces like this, `cudf:` is removed. What's the convention followed here?

File: cpp/include/cudf/utilities/error.hpp
Comment (karthikeyann) [inline cpp/include/cudf/utilities/error.hpp:240]: Could this be detected using compiler warnings?

File: cpp/tests/copying/gather_str_tests.cpp
Comment (srinivasyadav18) [inline cpp/tests/copying/gather_str_tests.cpp:128]: do we need `int&` ref here ? I think `int` or `const int` should be sufficient. 
There is an inconsistency in using range based for loop with this and the above loop i.e in `TEST_F(GatherTestStr, Gather)`

**Comment by PointKernel** [comment]
Question: do we enforce this linting in CI?

File: cpp/include/cudf/detail/structs/utilities.hpp
Comment (vyasr) [reply cpp/include/cudf/detail/structs/utilities.hpp:122]: Hmm this looks like the autofix didn't quite do what it intended. What it wants here is to also change the parameter to be pass-by-value, then moved (see [this rule](https://clang.llvm.org/extra/clang-tidy/checks/modernize/pass-by-value.html)). I'll apply it manually, good catch!

File: cpp/include/cudf/scalar/scalar.hpp
Comment (vyasr) [reply cpp/include/cudf/scalar/scalar.hpp:570]: It is coming from [this rule](https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-override.html). In this particular case (a virtual, no-op destructor) the arguments on that page don't really apply. However, it does help in one way: it ensures a compiler error if the parent class's destructor is not declared virtual (which is UB when you have inheritance).

File: cpp/include/cudf/scalar/scalar_factories.hpp
Comment (vyasr) [reply cpp/include/cudf/scalar/scalar_factories.hpp:81]: I am not sure why the namespacing changes happened. Let me do another pass where I fully qualify all the namespaces, then see if I can get clang-tidy to remove them and tell me why. Part of me suspects that this might have happened during some iterative runs of the autofixer but that in the current state it won't care about them.

File: cpp/include/cudf/utilities/error.hpp
Comment (vyasr) [reply cpp/include/cudf/utilities/error.hpp:240]: IIRC this can be detected with host compilers but not nvcc. I'd have to double-check that, though.

File: cpp/include/cudf/scalar/scalar_factories.hpp
Comment (vyasr) [reply cpp/include/cudf/scalar/scalar_factories.hpp:47]: @karthikeyann [asked the same question](https://github.com/rapidsai/cudf/pull/15894#discussion_r1625368654) :slightly_smiling_face:  Let's continue the discussion on that thread.

File: cpp/include/cudf/scalar/scalar_factories.hpp
Comment (vyasr) [reply cpp/include/cudf/scalar/scalar_factories.hpp]: OK, I tested this in https://github.com/rapidsai/cudf/pull/15894/commits/82b73a019f02b09f43c89f6045ca030db85ec2d3, and as I suspected the autofix did not change it again. Let's stick with fully qualified namespaces, I think we all generally prefer that.

File: cpp/tests/copying/gather_str_tests.cpp
Comment (vyasr) [reply cpp/tests/copying/gather_str_tests.cpp:128]: Good call, yes this is unnecessary. I'll remove the ref qualifier.

File: cpp/include/cudf/utilities/error.hpp
Comment (vyasr) [reply cpp/include/cudf/utilities/error.hpp:240]: OK, yes, this can be enforced with `-Wzero-as-null-pointer-constant`. However, that leads to warnings from `cuda_runtime_api.h`. I'm not sure if there is a good way to avoid that.

**Comment by vyasr** [comment]
> Question: do we enforce this linting in CI?

Not yet. The plan is to get things fixed first, then add a CI check. However, that may have to wait a few months, see https://github.com/rapidsai/cudf/pull/15894#issuecomment-2142801867.

**Review by srinivasyadav18 — COMMENTED** [review]
Changes looks good to me!

**Review by ttnghia — APPROVED** [review]
Java approval.

File: cpp/tests/copying/slice_tests.cpp
Comment (karthikeyann) [inline cpp/tests/copying/slice_tests.cpp:502]: nit: same comment about validity mask. readability is better, if the each data aligns with its validity too.

File: cpp/src/io/parquet/ipc/Schema_generated.h
Comment (karthikeyann) [inline cpp/src/io/parquet/ipc/Schema_generated.h:142]: This file has this comment on the top.
`// automatically generated by the FlatBuffers compiler, do not modify`
if it is possible to enforce east const with `FlatBuffers compiler`, please add that as comment in top of the file too.

**Comment by KyleFromNVIDIA** [comment]
https://github.com/rapidsai/cudf/pull/15894#issuecomment-2142728338

> which is the only way I can imagine filtering out a specific set of files right now.

CMake handles this by putting vendored files in their own directory and then adding a `.clang-tidy` file to that directory. For example: https://gitlab.kitware.com/cmake/cmake/-/blob/72607d3402b91c3a9f4579c86175e4ca1ce6aa4e/Utilities/.clang-tidy

**Comment by vyasr** [comment]
That could definitely work too, although we'd potentially need a separate such directory for tests and cudf source. At this point I'm more inclined to wait for clang 19 since it's around the corner, but if for some reason the tooling that provides isn't sufficient I'd be happy to see the reorg. Alternatively, if you're motivated to get this done sooner feel free to try your approach. I agree that it ought to work and it should be pretty easy.

**Pull request merged**
